plugins/policy-recommendation/policy_recommendation_job.py

```diff
@@ -0,0 +1,843 @@
+                        pod_selector = kubernetes.client.V1LabelSelector())]
+                )],
+                ingress = [antrea_crd.Rule(
+                    action = "Allow",
+                    _from = [antrea_crd.NetworkPolicyPeer(
+                        pod_selector = kubernetes.client.V1LabelSelector())]
+                )],
+            )
+        )
+        policies.append(dict_to_yaml(acnp.to_dict()))
+    return policies
+
+def generate_sql_query(table_name, limit, start_time, end_time, unprotected):
+    sql_query = "SELECT {} FROM {}".format(", ".join(FLOW_TABLE_COLUMNS), table_name)
+    if unprotected:
+        sql_query += " WHERE ingressNetworkPolicyName == '' AND egressNetworkPolicyName == ''"
+    else:
+        # Select user trusted denied flows when unprotected equals False
+        sql_query += " WHERE trusted == 1"
+    if start_time:
+        sql_query += " AND flowEndSeconds >= '{}'".format(start_time)
+    if end_time:
+        sql_query += " AND flowEndSeconds < '{}'".format(end_time)
+    sql_query += " GROUP BY {}".format(", ".join(FLOW_TABLE_COLUMNS))
+    if limit:
+        sql_query += " LIMIT {}".format(limit)
+    return sql_query
+
+def read_flow_df(spark, db_jdbc_address, sql_query, rm_labels):
+    flow_df = spark.read \
+        .format("jdbc") \
+        .option("driver", "ru.yandex.clickhouse.ClickHouseDriver") \
+        .option("url", db_jdbc_address) \
+        .option("user", os.getenv("CH_USERNAME")) \
+        .option("password", os.getenv("CH_PASSWORD")) \
+        .option("query",  sql_query)\
+        .load()
+    if rm_labels:
+        flow_df = flow_df.withColumn('sourcePodLabels', udf(remove_meaningless_labels, StringType())("sourcePodLabels"))\
+            .withColumn('destinationPodLabels', udf(remove_meaningless_labels, StringType())("destinationPodLabels"))\
+            .dropDuplicates(['sourcePodLabels', 'destinationPodLabels'])
+    flow_df = flow_df.withColumn('flowType', udf(get_flow_type, StringType())("destinationServicePortName", "destinationPodLabels"))
+    return flow_df
+
+def write_recommendation_result(spark, result, recommendation_type, db_jdbc_address, table_name, id):
+    if not id:
+        recommendation_id = str(uuid.uuid4())
+    else:
+        recommendation_id = id
+    result_dict = {
+        'id': recommendation_id,
+        'type': recommendation_type,
+        'timeCreated': datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
+        'yamls': '---\n'.join(filter(None, result))
+    }
+    result_df = spark.createDataFrame([result_dict])
+    result_df.write\
+        .mode("append") \
+        .format("jdbc") \
+        .option("driver", "ru.yandex.clickhouse.ClickHouseDriver") \
+        .option("url", db_jdbc_address) \
+        .option("user", os.getenv("CH_USERNAME")) \
+        .option("password", os.getenv("CH_PASSWORD")) \
+        .option("dbtable", table_name) \
+        .save()
+    return recommendation_id
```

Comment:
@yanjunz97  do you think we need to update the clickhouse monitor to periodically clean up recommendation results as well? Do you think perhaps we might need to define an expiration time for results, as perhaps it's not ok to bluntly delete reco results when memory exceeds threshold. 

In any case, I am ok not supporting periodic collelction of old reco results in Theia's first release.

Reply:
As we do not expect the recommendation results may occupy too many spaces, I think expiration time might be more reasonable comparing to cleaning up by the monitor. 

But I'm not sure what expiration time should be chosen. I think a recommendation policy might be useful for a long time. Maybe it is more reasonable to delete them only when users trigger a deletion task from the UI?

Reply:
Thanks @yanjunz97, that's valuable feedback. Nothing we need to address here, but we surely need a mechanism to handle lifecycle of policy recommendation results.